### PR TITLE
Prefix HAVE_ names of flint-config.h

### PR DESCRIPTION
**Pull request by jengelh**
A library such as flint should always try to namespace its own
    defines somehow, lest flint's defines could interfere or override a
    macro of the same name in a project that happens to use HAVE_* macros
    of its own.

### CODE REVIEW

**Comment by wbhart** [comment]
@jengelh Sorry to bother you. Could you rebase this on current trunk. We had a rather large PR that went in that screwed a few things up. But given the amount of work involved, it had to go first.

**Comment by jengelh** [comment]
updated.

**Comment by wbhart** [comment]
Thanks.

**Comment by wbhart** [comment]
Something seems to have gone wrong with alloca. Perhaps some necessary lines were removed to redefine alloca to _alloca on Windows under MSVC.

**Comment by tthsqe12** [comment]
So, I will put the blas matrix mul in after this is finished. Can we be sure that HAVE_BLAS (or FLINT_HAVE_BLAS) is working after this?

File: CMakeLists.txt [hunk 1]
```diff
@@ -112,64 +112,13 @@ with open(join('${CMAKE_SOURCE_DIR}','qadic', 'CPimport.txt')) as fin:
...
-
-if(NOT HAVE_BIG_ENDIAN)
-    set(HAVE_LITTLE_ENDIAN ON)
-    set(HAVE_DOUBLE_IEEE_LITTLE_ENDIAN ON)
-else()
-    set(HAVE_DOUBLE_IEEE_BIG_ENDIAN ON)
-endif()
-
 # setup for flint-config.h
-check_include_files(alloca.h HAVE_ALLOCA_H)
```
Comment (dimpase) [inline CMakeLists.txt:126]: this has caused errors on MS's cl compiler

File: CMakeLists.txt [hunk 2]
```diff
@@ -112,64 +112,13 @@ with open(join('${CMAKE_SOURCE_DIR}','qadic', 'CPimport.txt')) as fin:
...
-check_include_files(sys/systemcfg.h HAVE_SYS_SYSTEMCFG_H)
-check_include_files(sys/times.h HAVE_SYS_TIMES_H)
-check_include_files(sys/time.h HAVE_SYS_TIME_H)
-check_include_files(sys/types.h HAVE_SYS_TYPES_H)
-check_include_files(unistd.h HAVE_UNISTD_H)
-check_include_files(dlfcn.h HAVE_DLFCN_H)
-check_include_files(fpu_control.h HAVE_FPU_CONTROL_H)
-
-
-check_function_exists(alloca HAVE_ALLOCA)
```
Comment (dimpase) [inline CMakeLists.txt:153]: On MS cl, aka MS Visual C, `alloca` is called `_alloca`, and its prototype is in `malloc.h`.

File: config.h.in [hunk 3]
```diff
@@ -1,329 +1,16 @@
...
-#define HAVE_UINTMAX_T       1
-#define HAVE_PTRDIFF_T       1
-#define HAVE_UINT_LEAST32_T  1
-#define SIZEOF_UINTMAX_T	 8
-#endif
-
-#define NPRINTF_J            1
-#define NPRINTF_T            1
-
-#ifdef _MSC_VER
```
Comment (dimpase) [inline config.h.in:316]: the removal of this `#ifdef` section looks dodgy. E.g. definition of `alloca` is certainly needed. Perhaps cmake has a better way to handle this, @isuruf might know, as I imagine it's a common portability issue.

Also, perhaps `_alloca` should be `_malloca`, as https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/alloca?view=vs-2019 seems to recommend - but that's another story.

**Review by dimpase — COMMENTED** [review]
this definitely breaks MS `cl` build, as `alloca` handling is gone.

File: config.h.in (hunk 3 above)
Comment (jengelh) [reply config.h.in:316]: Restored that part.

**Comment by dimpase** [comment]
I suppose one has to rebase the PR over the current develop, so that it's mergeable again, otherwise CI fails.

**Comment by tthsqe12** [comment]
yay! @wbhart

**Comment by wbhart** [comment]
Thank you indeed!

**Comment by mahrud** [comment]
`FLINT_WANT_ASSERT` and `FLINT_USES_PTHREAD` are never defined in CMakeLists.txt. The former is not absolutely necessary, but without the latter mutexes are never used in threaded routines. In particular, programs that defined `HAVE_PTHREAD` and include flint may be buggy now. Unless I'm not understanding something, shouldn't at least some ci tests fail because of this?

**Comment by jengelh** [comment]
>FLINT_WANT_ASSERT and FLINT_USES_PTHREAD are never defined in CMakeLists.txt

Their previous names were not defined in the previous git state's cmake system, either.

>programs that defined HAVE_PTHREAD and include flint may be buggy now

Quite the contrary, they were buggy and (reasonably) fixed now. A program setting HAVE_PTHREAD for itself would previously would get a mismatch of the libflint ABI, because libflint may have been built with a different HAVE_PTHREAD value.

**Comment by mahrud** [comment]
> Their previous names were not defined in the previous git state's cmake system, either.

The way CMake works, if you build flint it from a parent project, variables defined will be passed onto CMake, so if `HAVE_PTHREAD` is defined in the CMake files of the parent project, it'll be also defined for flint, hence `#cmakedefine HAVE_PTHREAD 1` will be converted to `#define HAVE_PTHREAD 1`, but now this doesn't happen.

I'm not saying that the previous behavior (where the previous names were not defined by flint) was correct, but at least it worked because different projects use roughly the same names. Now that you've changed the names, you should also define them in CMakeLists.txt.

**Comment by dimpase** [comment]
> variables defined by the parent project's CMake files

Do you mean by "variables" as something only CMake understands, i.e. a CMake configuration feature?
(CMake made a ridiculously confusing  choice of `-D` syntax for passing parameters, such as `-DWITH_NTL=yes/no` in Flint, something that makes one think that  it's merely a value of a preprocessor variable `WITH_NTL` being passed to Flint build system.)

If it's the case, then the parent project somehow assumes that Flint's requirements on pthreads are same (or weaker) than these of the parent' Something that's not guaranteed.

If on the other hand if that variable is merely #define'd in a header of a parent project, then Flint may or may not be aware of its presence, and it's up to Flint what to do with it, if anything. Such a #define typically goes in an internal header, not normally #included by anybody but the project itself.

What one probably really needs here is a Flint CMake config parameter that controls whether Flint is built with pthreads, something akin to `-DWITH_NTL`.

**Pull request merged**